compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertAIEWorkgroup.cpp

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEDialect.h"
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/AMDAIEOpUtils.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree/compiler/Codegen/TransformStrategies/GPU/Common.h"
+
+#define DEBUG_TYPE "iree-amdaie-insert-aie-workgroup"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// Utility to map the parallel mapping attributes to the corresponding
+/// induction variables.
+void getAttributeMapping(SmallVector<scf::ForallOp> forallOps,
+                         DenseMap<Attribute, Value> &attrMapping) {
+  for (auto forallOp : forallOps) {
+    if (!forallOp.getMapping().has_value()) continue;
+    SmallVector<Attribute> mapping =
+        llvm::to_vector(forallOp.getMapping()->getValue());
+    auto ivs = forallOp.getInductionVars();
+    for (auto &&[attr, iv] : llvm::zip(mapping, ivs)) attrMapping[attr] = iv;
+  }
+}
+
+/// Insert workgroups around `scf::Forall` ops with thread mapping.
+/// NOTE: This can be extended to block mapping as well later.
+LogicalResult insertWorkgroupOps(mlir::ModuleOp moduleOp) {
+  IRRewriter rewriter(moduleOp.getContext());
+  moduleOp->walk([&](scf::ForallOp forallOp) {
+    if (forallOp.getNumResults() != 0) {
+      return WalkResult::advance();
+    }
+    if (!forallOp.getMapping().has_value()) {
+      return WalkResult::advance();
+    }
+    SmallVector<Attribute> threadMapping =
+        llvm::to_vector(forallOp.getMapping()->getValue());
+    if (llvm::any_of(threadMapping, [](Attribute map) {
+          return !llvm::isa<mlir::gpu::GPUThreadMappingAttr>(map);
+        })) {
+      return WalkResult::advance();
+    }
+    if (threadMapping.size() != 2) {
+      return WalkResult::advance();
+    }
+    // Create the workgroup region
+    rewriter.setInsertionPoint(forallOp);
+    auto workgroupOp =
+        rewriter.create<AMDAIE::WorkgroupOp>(rewriter.getUnknownLoc());
+    rewriter.moveOpBefore(forallOp, workgroupOp.getControlCode());
+    return WalkResult::advance();
+  });
+  return success();
+}
+
+/// Insert core ops inside innermost forall ops around computational ops and
+/// add synchronization ops along the way to synchronize with surrounding
+/// dma ops.
+LogicalResult insertCoreOpsInWorkgroup(mlir::ModuleOp moduleOp) {
+  IRRewriter rewriter(moduleOp.getContext());
+  WalkResult res = moduleOp->walk([&](AMDAIE::WorkgroupOp workgroupOp) {
+    // Find innermost forall loops
+    SmallVector<scf::ForallOp> innermostForallLoops =
+        getInnermostForallLoops(workgroupOp);
+    if (innermostForallLoops.empty()) {
+      workgroupOp.emitOpError() << "contains no nested forall op";
+      return WalkResult::interrupt();
+    }
+    if (innermostForallLoops.size() != 1) {
+      workgroupOp.emitOpError()
+          << "multiple innermost foralls not supported for now";
+      return WalkResult::interrupt();
+    }
```

Comment:
For peeled codes, there will be multiple innermost foralls. Do you have plans to add support in the following PRs?

Reply:
Peeling is already supported as the workgroup will end up around every `forall`, so in case of peeling there will be multiple workgroups.

Reply:
I updated the test to multiple innermost `forall` ops so you can see what the output looks like.